contracts/SimpleSaleContract.sol

```diff
@@ -16,47 +23,140 @@ contract SimpleSaleContract is Ownable {
+    // is it better to specity the default visibility modifier public or omit it?
+    // OMG because it doesn't have the public keyword its getter is absent in the abi!!!
+    mapping(IERC20 => ExchangeRate) public paymentTokensExchangeRate;
+    ExchangeRate public etherExchangeRate;
 
-    event SaleCompleted(address indexed beneficiary, uint256 tokensBought, address indexed paymentToken, uint256 tokensPaid);
-    event TokenExchangeRateSet(address indexed token, uint256 exchangeRate);
+    event TokenSaleCompleted(address indexed beneficiary, uint256 tokensBought, IERC20 indexed paymentToken, uint256 tokensPaid);
+    event EtherSaleCompleted(address indexed beneficiary, uint256 tokensBought, uint256 etherPaid);
+    event TokenExchangeRateSet(IERC20 indexed token, uint256 partsSellToken, uint256 partsMintToken);
+    event EtherExchangeRateSet(uint256 partsSellEther, uint256 partsMintToken);
+    event TokensWithdrawn(IERC20 indexed token, uint256 value);
+
+    error BadEthValue();
 
     /**
      * @notice contract constructor
      * @param _mintableToken The token that will be minted by this contract. This contract must have the rights to mint.
+     * @param _saleDurationDays The period in days after the contract deployment during which it is possible to sell tokens
      */
-    constructor(address _mintableToken) {
+    constructor(address _mintableToken, uint256 _saleDurationDays) {
         mintableToken = MeetupERC20Token(_mintableToken);
+        cutoffTimestamp = block.timestamp + _saleDurationDays * 1 days;
+    }
+
+    modifier rateSet(IERC20 token) {
+        require(paymentTokensExchangeRate[token].partsSell != 0, "Payment token not allowed");
+        require(paymentTokensExchangeRate[token].partsMint != 0, "Payment token not allowed");
+        _;
+    }
+
+    modifier forSale() {
+        require(block.timestamp < cutoffTimestamp, "Sale is not possible anymore");
+        _;
+    }
+
+    modifier positiveTokens(uint256 value) {
+        require(value != 0, "Tokens value should be positive");
+        _;
     }
 
     /**
-     * @notice Transfers mintableTokens to msg.sender for a specified exchange rate. Reverts if payment token is not specified.
+     * @notice In exchange for _paymentToken, mintableTokens will be transferred to msg.sender at a specified exchange rate.
+     * Reverts if payment token is not specified.
      * @param _tokensToBuy amount of mintable tokens to be bought
      * @param _paymentToken the token that will be used for payment
      */
-    function buyTokens(uint256 _tokensToBuy, address _paymentToken) external {
-        require(_tokensToBuy > 0, "No tokens bought");
-        require(paymentTokensExchangeRate[IERC20(_paymentToken)] > 0, "Payment token not allowed");
+    function buyTokens(uint256 _tokensToBuy, IERC20 _paymentToken) external forSale positiveTokens(_tokensToBuy) rateSet(_paymentToken) {
+        require(_tokensToBuy != 0, "No tokens bought");
```

Comment:
You are using a modifier `positiveTokens` and still you check the same in the function body